<a href="https://colab.research.google.com/github/wajidalianalytics/Data-Analytics-Using-Python/blob/main/Lecture3_Data_Structures_Starbucks_Sample.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lecture 3: Data Structures in Python
### Data Analytics using Python

## Our Business Case: Starbucks India ☕
Throughout this lecture we use **one dataset**: `Starbucks_India_Sales_Dataset.xlsx`
(100 customer orders from July to September 2026, 15 variables, 5 cities, 10 stores, 8 products).
Every data structure we learn is applied to this same Excel data.

> Note: This is **synthetic (made-up) data** for teaching, not real Starbucks data.

### The Dataset: 15 Variables
| # | Variable | Type | Meaning |
|---|---|---|---|
| 1 | `OrderID` | Text | Unique order number (SBX1001 ...) |
| 2 | `Date` | Date | Order date (Jul-Sep 2026) |
| 3 | `City` | Category | Delhi, Mumbai, Bengaluru, Hyderabad, Pune |
| 4 | `StoreID` | Text | Store code (S01 ... S10) |
| 5 | `Product` | Category | Cappuccino, Latte, Cold Brew, ... |
| 6 | `Category` | Category | Hot Coffee, Cold Coffee, Tea, Food |
| 7 | `Size` | Category | Tall, Grande, Venti, Regular |
| 8 | `Unit_Price` | Number (Rs) | Price of one item |
| 9 | `Quantity` | Number | Items in the order |
| 10 | `Discount_Pct` | Number (%) | Discount given (0 to 20%) |
| 11 | `Revenue` | Number (Rs) | `Unit_Price x Quantity x (1 - Discount_Pct)` |
| 12 | `Payment_Mode` | Category | UPI, Card, Cash, Wallet |
| 13 | `Customer_Type` | Category | New, Returning, Member |
| 14 | `Order_Channel` | Category | In-store, App, Delivery |
| 15 | `Rating` | Number (1-5) | Customer rating (3 values are blank on purpose) |

The Excel file has **4 sheets**: `Sales_Data` (main table), `Products` and `Stores` (master tables), `Data_Dictionary`.

### Lecture Plan
| Section | Topic | Time |
|---|---|---|
| 1 | Create data structures (list, tuple, set, dict)
| 2 | Access elements
| 3 | Functions on data structures
| 4 | Which structure to use when?
| 5 | NumPy arrays
| 6 | Pandas Series and DataFrame
| 7 | Indexing and selection
| 8 | Organising business datasets

---
# Step 0: Load the Excel Data (run this first)

Put the Excel file in the **same folder** as this notebook. In **Google Colab**: click the 📁 folder icon on the left → upload `Starbucks_India_Sales_Dataset.xlsx`.

Do not worry about the loading code now. We study Pandas properly in Section 6. For now, it simply reads each Excel sheet into a table called a **DataFrame**.

| Term | Meaning |
|---|---|
| `pd.read_excel()` | Reads an Excel sheet into Python |
| `df` | Short name for the main sales table (DataFrame) |
| `NaN` | "Not a Number": how Python shows a **missing value** (blank cell) |

In [2]:

import pandas as pd

# Load the Excel files
df = pd.read_excel(FILE, sheet_name="Sales_Data")
products_df = pd.read_excel(FILE, sheet_name="Products")
stores_df = pd.read_excel(FILE, sheet_name="Stores")

# Check the data
print("Sales table:", df.shape)
print("Products:", products_df.shape)
print("Stores:", stores_df.shape)

df.head()

Sales table: (100, 15)
Products: (8, 4)
Stores: (10, 5)


,OrderID,Date,City,StoreID,Product,Category,Size,Unit_Price,Quantity,Discount_Pct,Revenue,Payment_Mode,Customer_Type,Order_Channel,Rating
0,SBX1001,2026-07-01,Hyderabad,S07,Cappuccino,Hot Coffee,Grande,280,1,0.00,280.0,UPI,New,In-store,4.0
1,SBX1002,2026-07-01,Bengaluru,S05,Cold Brew,Cold Coffee,Tall,300,1,0.10,270.0,UPI,Member,In-store,4.0
2,SBX1003,2026-07-01,Mumbai,S03,Cappuccino,Hot Coffee,Venti,310,1,0.05,294.5,UPI,Returning,In-store,2.0
3,SBX1004,2026-07-02,Hyderabad,S08,Blueberry Muffin,Food,Regular,180,1,0.20,144.0,Wallet,Member,In-store,3.0
4,SBX1005,2026-07-04,Delhi,S01,Latte,Hot Coffee,Grande,300,2,0.20,480.0,Card,Member,In-store,4.0


---
# Section 1: Create Different Data Structures in Python

A **data structure** is a way of storing data so that we can use it easily.
Python has 4 built-in structures:

| Structure | Symbol | Ordered? | Changeable? | Duplicates? | Starbucks use |
|---|---|---|---|---|---|
| **List** | `[ ]` | Yes | Yes | Allowed | Menu items that can change |
| **Tuple** | `( )` | Yes | No | Allowed | A store's fixed details |
| **Set** | `{ }` | No | Yes | Not allowed | Unique cities |
| **Dictionary** | `{k: v}` | Yes | Yes | Keys unique | Product → Price |

### 📖 Definitions
| Term | Definition | Starbucks example |
|---|---|---|
| **Data structure** | A way of organising and storing data in memory so it can be used efficiently. | Storing menu, prices, cities |
| **List** | An *ordered* and *changeable* collection of items in square brackets `[ ]`. Duplicates allowed. | `["Latte", "Cold Brew"]` |
| **Tuple** | An *ordered* but *unchangeable* collection in round brackets `( )`. | `("S01", "Connaught Place", "Delhi")` |
| **Set** | An *unordered* collection of **unique** items in curly brackets `{ }`. | `{"Delhi", "Mumbai"}` |
| **Dictionary** | A collection of **key : value** pairs. Each key is unique and is used to find its value. | `{"Latte": 270}` |
| **Mutable** | Can be changed after creation (list, set, dict). | Add a new drink to the menu |
| **Immutable** | Cannot be changed after creation (tuple, string, number). | A store's fixed record |
| **Record** | One complete row of data, often stored as a dictionary. | One order with all its details |

In [3]:
# ---- Part A: create them by typing ----

# LIST: the Starbucks menu (8 products)
menu = ["Cappuccino", "Latte", "Americano", "Cold Brew",
        "Frappuccino", "Masala Chai", "Blueberry Muffin", "Veg Sandwich"]

# TUPLE: one store's fixed details (StoreID, name, city)
store = ("S01", "Connaught Place", "Delhi")

# SET: cities where Starbucks sells (the duplicate 'Delhi' is removed automatically)
cities = {"Delhi", "Mumbai", "Bengaluru", "Hyderabad", "Pune", "Delhi"}

# DICTIONARY: product -> base price in Rs
price = {"Cappuccino": 250, "Latte": 270, "Americano": 220, "Cold Brew": 300,
         "Frappuccino": 350, "Masala Chai": 200, "Blueberry Muffin": 180, "Veg Sandwich": 240}

print("List      :", menu)
print("Tuple     :", store)
print("Set       :", cities, " <- 'Delhi' typed twice, stored once")
print("Dictionary:", price)
print()
print(type(menu), type(store), type(cities), type(price))

List      : ['Cappuccino', 'Latte', 'Americano', 'Cold Brew', 'Frappuccino', 'Masala Chai', 'Blueberry Muffin', 'Veg Sandwich']
Tuple     : ('S01', 'Connaught Place', 'Delhi')
Set       : {'Bengaluru', 'Pune', 'Delhi', 'Hyderabad', 'Mumbai'}  <- 'Delhi' typed twice, stored once
Dictionary: {'Cappuccino': 250, 'Latte': 270, 'Americano': 220, 'Cold Brew': 300, 'Frappuccino': 350, 'Masala Chai': 200, 'Blueberry Muffin': 180, 'Veg Sandwich': 240}

<class 'list'> <class 'tuple'> <class 'set'> <class 'dict'>


In [4]:
# ---- Part B: create them from the Excel data ----
# Each Excel column becomes a Python list. Each Excel row becomes a dictionary.

product_list = df["Product"].tolist()        # 100 products (with repeats)
city_list    = df["City"].tolist()           # 100 cities  (with repeats)
qty_list     = df["Quantity"].tolist()       # 100 quantities
revenue_list = df["Revenue"].tolist()        # 100 revenue values

first_order   = df.iloc[0].to_dict()                 # ONE order  -> dictionary
first_5_orders = df.head(5).to_dict("records")       # FIVE orders -> list of dictionaries

print("Items in product_list :", len(product_list))
print("Unique cities (set)   :", set(city_list))
print()
print("First order as a dictionary:")
print(first_order)

Items in product_list : 100
Unique cities (set)   : {'Pune', 'Hyderabad', 'Bengaluru', 'Delhi', 'Mumbai'}

First order as a dictionary:
{'OrderID': 'SBX1001', 'Date': Timestamp('2026-07-01 00:00:00'), 'City': 'Hyderabad', 'StoreID': 'S07', 'Product': 'Cappuccino', 'Category': 'Hot Coffee', 'Size': 'Grande', 'Unit_Price': 280, 'Quantity': 1, 'Discount_Pct': 0.0, 'Revenue': 280.0, 'Payment_Mode': 'UPI', 'Customer_Type': 'New', 'Order_Channel': 'In-store', 'Rating': 4.0}


---
# Section 2: Access Elements of Data Structures (6 min)

- **List / Tuple**: use the **position (index)**. Python starts counting from **0**.
- **Dictionary**: use the **key**.
- **Set**: has no position, so we check *membership* (`in`) or loop over it.

```
Index:    0           1        2          3            ...   7
menu = ["Cappuccino","Latte","Americano","Cold Brew", ... ,"Veg Sandwich"]
Negative: -8          -7       -6         -5           ...   -1
```

### 📖 Definitions
| Term | Definition | Starbucks example |
|---|---|---|
| **Element / Item** | One single value inside a data structure. | `"Latte"` in the menu |
| **Index** | The position number of an element. Starts from **0**. Negative index counts from the end (-1 = last). | `menu[0]` = Cappuccino |
| **Slicing** | Taking a part of a sequence using `[start:stop:step]`. The stop position is *not* included. | `menu[0:3]` = first three items |
| **Key** | The unique label used to find a value in a dictionary. | `price["Latte"]` |
| **Nested structure** | A structure inside another structure, e.g. a list of dictionaries. | `first_5_orders[2]["City"]` |
| **Membership test** | Checking if a value exists using `in`. Returns `True` or `False`. | `"Pune" in cities` |
| **`.get()`** | Dictionary method that returns a default value instead of an error when the key is missing. | `price.get("Pizza", "Not on menu")` |

In [4]:
# --- List ---
print("First item      :", menu[0])
print("Last item       :", menu[-1])
print("First three     :", menu[0:3])      # slicing: start included, end excluded
print("Every 2nd item  :", menu[::2])

# --- Tuple (same as list) ---
print("\nStore ID        :", store[0])
print("Store city      :", store[2])

# --- Dictionary ---
print("\nPrice of Latte  :", price["Latte"])
print("Safe lookup     :", price.get("Pizza", "Not on menu"))   # no error if key is missing

# --- Set ---
print("\nIs Pune a Starbucks city?   ", "Pune" in cities)
print("Is Jaipur a Starbucks city? ", "Jaipur" in cities)

First item      : Cappuccino
Last item       : Veg Sandwich
First three     : ['Cappuccino', 'Latte', 'Americano']
Every 2nd item  : ['Cappuccino', 'Americano', 'Frappuccino', 'Blueberry Muffin']

Store ID        : S01
Store city      : Delhi

Price of Latte  : 270
Safe lookup     : Not on menu

Is Pune a Starbucks city?    True
Is Jaipur a Starbucks city?  False


In [5]:
# --- Changing data: a list is changeable, a tuple is not ---
menu[6] = "Muffin"
print("Updated menu :", menu)
menu[6] = "Blueberry Muffin"          # change it back, we need the original name later

try:
    store[1] = "Saket Mall"
except TypeError as e:
    print("Tuple error  :", e)

Updated menu : ['Cappuccino', 'Latte', 'Americano', 'Cold Brew', 'Frappuccino', 'Masala Chai', 'Muffin', 'Veg Sandwich']
Tuple error  : 'tuple' object does not support item assignment


In [6]:
# --- Access the real Excel data stored in Python structures ---
print("City of order #1       :", city_list[0])
print("Quantity of last order :", qty_list[-1])
print("First 5 revenues       :", revenue_list[:5])

print("\nProduct of first order :", first_order["Product"])
print("Revenue of first order :", first_order["Revenue"])

# list of dictionaries: [position][key]
print("\nCity of 3rd order      :", first_5_orders[2]["City"])
print("Product of 3rd order   :", first_5_orders[2]["Product"])

City of order #1       : Hyderabad
Quantity of last order : 2
First 5 revenues       : [280.0, 270.0, 294.5, 144.0, 480.0]

Product of first order : Cappuccino
Revenue of first order : 280.0

City of 3rd order      : Mumbai
Product of 3rd order   : Cappuccino


---
# Section 3: Apply Functions to Data Structures (7 min)

Python gives us ready-made **functions** (`len`, `sum`, `max`, `min`, `sorted`) and **methods** (`.append()`, `.keys()` ...) so we don't write everything ourselves.

### 📖 Definitions
| Term | Definition | Starbucks example |
|---|---|---|
| **Function** | A reusable block of code that takes input and gives output. Called as `name(data)`. | `sum(revenue_list)` |
| **Method** | A function that belongs to an object. Called as `object.name()`. | `menu.append("Iced Tea")` |
| **`len()`** | Counts the number of elements. | Number of orders |
| **`sum()` / `max()` / `min()`** | Total / largest / smallest value. | Total revenue, biggest order |
| **`sorted()`** | Returns a new list arranged in order. | Top 5 orders |
| **Loop (`for`)** | Repeats an action for every element. | Count orders in each city |
| **`zip()`** | Pairs elements of two sequences together. | Product with its revenue |
| **List comprehension** | A short one-line way to build a new list from an existing one. | `[r for r in revenue_list if r > 800]` |
| **Union `\|`** | All items from both sets. | Cities selling Cold Brew or Chai |
| **Intersection `&`** | Items present in **both** sets. | Cities selling both products |
| **Difference `-`** | Items in the first set but not in the second. | Cities with Cold Brew but no Chai |

In [7]:
# Built-in functions on the Revenue list (100 orders)
print("Number of orders   :", len(revenue_list))
print("Total revenue      : Rs", round(sum(revenue_list), 2))
print("Average order value: Rs", round(sum(revenue_list) / len(revenue_list), 2))
print("Biggest order      : Rs", max(revenue_list))
print("Smallest order     : Rs", min(revenue_list))
print("Top 5 orders       :", sorted(revenue_list, reverse=True)[:5])
print("Total cups/items   :", sum(qty_list))

Number of orders   : 100
Total revenue      : Rs 43652.5
Average order value: Rs 436.52
Biggest order      : Rs 2050.0
Smallest order     : Rs 144.0
Top 5 orders       : [2050.0, 1282.5, 990.0, 930.0, 900.0]
Total cups/items   : 173


In [14]:
# Count orders in each city

orders_per_city = {}

for c in city_list:

    if c in orders_per_city:
        orders_per_city[c] = orders_per_city[c] + 1
    else:
        orders_per_city[c] = 1

print("Orders per city:", orders_per_city)


# Find the busiest city
busiest_city = ""
highest_orders = 0

for c, orders in orders_per_city.items():

    if orders > highest_orders:
        highest_orders = orders
        busiest_city = c

print("Busiest city:", busiest_city)

Orders per city: {'Hyderabad': 20, 'Bengaluru': 25, 'Mumbai': 25, 'Delhi': 16, 'Pune': 14}
Busiest city: Bengaluru


In [12]:
# Total revenue for each product
revenue_by_product = {}

for p, r in zip(product_list, revenue_list):

    if p in revenue_by_product:
        revenue_by_product[p] = revenue_by_product[p] + r
    else:
        revenue_by_product[p] = r

for p, r in revenue_by_product.items():
    print(p, "-> Rs", r)

# Find the product with the highest revenue
best_product = ""
highest_revenue = 0

for p, r in revenue_by_product.items():
    if r > highest_revenue:
        highest_revenue = r
        best_product = p

print("Best-selling product by revenue:", best_product)

Cappuccino -> Rs 5289.0
Cold Brew -> Rs 6829.5
Blueberry Muffin -> Rs 4293.0
Latte -> Rs 8938.5
Veg Sandwich -> Rs 5136.0
Americano -> Rs 4273.0
Frappuccino -> Rs 6101.5
Masala Chai -> Rs 2792.0
Best-selling product by revenue: Latte


In [15]:
# Find orders above Rs 800
high_value = []

for r in revenue_list:
    if r > 800:
        high_value.append(r)

print("Orders above Rs 800:", len(high_value), "->", high_value)


# Find products priced above Rs 260
premium = []

for p in price:
    if price[p] > 260:
        premium.append(p)

print("Premium products:", premium)


# Add a product to the menu
menu.append("Iced Tea")

# Remove the product from the menu
menu.remove("Iced Tea")

# Count the number of products in the menu
print("\nMenu size:", len(menu))

# Show the first 3 product names from the dictionary
print("Keys:", list(price.keys())[:3])

# Show the first 3 prices from the dictionary
print("Values:", list(price.values())[:3])

Orders above Rs 800: 9 -> [836.0, 1282.5, 900.0, 2050.0, 810.0, 900.0, 930.0, 990.0, 820.0]
Premium products: ['Latte', 'Cold Brew', 'Frappuccino']

Menu size: 8
Keys: ['Cappuccino', 'Latte', 'Americano']
Values: [250, 270, 220]


In [17]:
# Find cities where Cold Brew was sold
cold_brew_cities = set()

for c, p in zip(city_list, product_list):
    if p == "Cold Brew":
        cold_brew_cities.add(c)

# Find cities where Masala Chai was sold
chai_cities = set()

for c, p in zip(city_list, product_list):
    if p == "Masala Chai":
        chai_cities.add(c)

print("Cold Brew sold in:", cold_brew_cities)
print("Masala Chai sold in:", chai_cities)

# Cities where both products were sold
print("Both products:", cold_brew_cities & chai_cities)

# Cities where either product was sold
print("Any of the two:", cold_brew_cities | chai_cities)

# Cities where Cold Brew was sold but Masala Chai was not
print("Cold Brew but no Chai:", cold_brew_cities - chai_cities)

Cold Brew sold in: {'Mumbai', 'Hyderabad', 'Bengaluru', 'Delhi', 'Pune'}
Masala Chai sold in: {'Bengaluru', 'Mumbai', 'Pune', 'Delhi'}
Both products: {'Bengaluru', 'Delhi', 'Pune', 'Mumbai'}
Any of the two: {'Mumbai', 'Hyderabad', 'Bengaluru', 'Delhi', 'Pune'}
Cold Brew but no Chai: {'Hyderabad'}


---
# Section 4: Evaluate the Utility of Data Structures (6 min)

**Which one should a Starbucks analyst choose?**

| Business situation (in our dataset) | Best choice | Why |
|---|---|---|
| Menu that changes often | **List** | Easy to add, remove, change |
| A store's fixed record (ID, name, city) | **Tuple** | Cannot be changed by mistake |
| Unique cities / payment modes | **Set** | No duplicates, very fast search |
| Product → price, or OrderID → full order | **Dictionary** | Instant lookup by name/ID |
| 100,000 prices for fast maths | **NumPy array** | Very fast calculations |
| The whole 100 × 15 sales table | **Pandas DataFrame** | Best for business analysis |

### 📖 Definitions
| Term | Definition | Starbucks example |
|---|---|---|
| **Utility** | How useful and suitable a data structure is for a given task. | Using a set to find unique payment modes |
| **Lookup / Search** | Finding whether a value exists or fetching its value. | Find order SBX1050 |
| **Duplicate** | The same value appearing more than once. | 'Delhi' appears many times in `city_list` |
| **Time efficiency** | How fast an operation runs. Sets and dictionaries search much faster than lists for large data. | Searching 1 million IDs |
| **Trade-off** | Every structure is good at something and weak at something else. Choose based on the task. | List keeps order, set removes duplicates |

In [18]:
# Q1. How many DIFFERENT payment modes and customer types do we have?  -> SET
payment_list  = df["Payment_Mode"].tolist()
customer_list = df["Customer_Type"].tolist()

print("Payment entries :", len(payment_list), " | Unique modes   :", set(payment_list))
print("Customer entries:", len(customer_list), " | Unique types   :", set(customer_list))

Payment entries : 100  | Unique modes   : {'Cash', 'UPI', 'Card', 'Wallet'}
Customer entries: 100  | Unique types   : {'New', 'Member', 'Returning'}


In [21]:
# Q2. Find the full details of order SBX1050.  -> DICTIONARY (lookup by key)
orders_dict = {}

for o in df.to_dict("records"):
    orders_dict[o["OrderID"]] = o

o = orders_dict["SBX1050"]

print("Product:", o["Product"])
print("City:", o["City"])
print("Revenue:", o["Revenue"])

Product: Frappuccino
City: Delhi
Revenue: 342.0


In [22]:
o = orders_dict["SBX1005"]

print("Order SBX1005")
print("Product:", o["Product"])
print("City:", o["City"])
print("Revenue:", o["Revenue"])

Order SBX1005
Product: Latte
City: Delhi
Revenue: 480.0


---
# Section 5: NumPy Arrays (8 min)

**NumPy** stores numbers in an **array**: like a list, but much faster, and it can do maths on **all values at once** (no loop needed).

We now move the numeric columns of our Excel data (`Unit_Price`, `Quantity`, `Discount_Pct`, `Revenue`) into NumPy arrays.

### 📖 Definitions
| Term | Definition | Starbucks example |
|---|---|---|
| **NumPy** | *Numerical Python*: a library for fast mathematical operations on numbers. | `import numpy as np` |
| **Array (`ndarray`)** | A grid of values of the **same data type**, stored compactly for speed. | `unit_price` array |
| **1D / 2D array** | 1D = one row of values (like a list). 2D = rows and columns (like a table). | 2D: orders × 4 numeric columns |
| **Shape** | The size of the array as (rows, columns). | `(100, 4)` |
| **`ndim`** | Number of dimensions of the array. | 2 |
| **`dtype`** | The data type of the values inside the array. | `int64`, `float64` |
| **Vectorisation** | Doing maths on the whole array at once without a loop. | `unit_price * qty * (1 - discount)` |
| **Axis** | Direction of calculation. `axis=0` goes down the rows (per column). `axis=1` goes across (per row). | Average of each column |
| **Aggregation** | Reducing many values to one summary value. | `sum`, `mean`, `max` |
| **Boolean mask** | An array of True/False created from a condition, used to filter or count. | `qty >= 3` |
| **`NaN`** | "Not a Number": a missing value. `np.nanmean` ignores it. | 3 blank Ratings |

In [25]:
# 1D arrays: one array per numeric column.
order_ids  = df["OrderID"].to_numpy()
unit_price = df["Unit_Price"].to_numpy()
qty        = df["Quantity"].to_numpy()
discount   = df["Discount_Pct"].to_numpy()

print("Shape :", unit_price.shape, "-> 100 values in one dimension")
print("Type  :", unit_price.dtype)
print("First 5 prices   :", unit_price[:5])
print("First 5 quantity :", qty[:5])
print("First 5 discount :", discount[:5])

Shape : (100,) -> 100 values in one dimension
Type  : int64
First 5 prices   : [280 300 310 180 300]
First 5 quantity : [1 1 1 1 2]
First 5 discount : [0.   0.1  0.05 0.2  0.2 ]


In [26]:
# Maths on the whole array at once (no loop). Rebuild the Revenue formula from Excel:it calculates the revenue for every order using NumPy, and second, it calculates some summary statistics such as total, average, median, and maximum revenue.
#it calculates the revenue for every order using NumPy,
#and second, it calculates some summary statistics such as total, average, median, and maximum revenue.

revenue_np = unit_price * qty * (1 - discount)

print("First 5 revenue (NumPy):", revenue_np[:5])
print("Same as the Excel Revenue column?", np.allclose(revenue_np, df["Revenue"].to_numpy()))

# Aggregations
print("\nTotal revenue  : Rs", revenue_np.sum().round(2))
print("Average        : Rs", revenue_np.mean().round(2))
print("Median         : Rs", np.median(revenue_np))
print("Std deviation  : Rs", revenue_np.std().round(2))
print("Max / Min      :", revenue_np.max(), "/", revenue_np.min())

i = revenue_np.argmax()                       # position of the maximum
print("\nBiggest order is", order_ids[i], "with Rs", revenue_np[i])

First 5 revenue (NumPy): [280.  270.  294.5 144.  480. ]
Same as the Excel Revenue column? True

Total revenue  : Rs 43652.5
Average        : Rs 436.52
Median         : Rs 351.0
Std deviation  : Rs 278.23
Max / Min      : 2050.0 / 144.0

Biggest order is SBX1052 with Rs 2050.0


In [28]:
# Find orders with 3 or more items

bulk_orders = 0
bulk_revenue = 0

for i in range(len(qty)):

    if qty[i] >= 3:
        bulk_orders = bulk_orders + 1
        bulk_revenue = bulk_revenue + revenue_np[i]

print("Orders with 3+ items:", bulk_orders)
print("Revenue from bulk orders: Rs", round(bulk_revenue, 2))


# Label each order as Bulk or Regular

order_type = []

for q in qty:

    if q >= 3:
        order_type.append("Bulk")
    else:
        order_type.append("Regular")

print("Order types:", order_type)


# Find the highest 5 revenues

top5_revenue = sorted(revenue_np, reverse=True)[:5]

print("Top 5 revenues:", top5_revenue)

Orders with 3+ items: 18
Revenue from bulk orders: Rs 14982.5
Order types: ['Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Bulk', 'Bulk', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Bulk', 'Bulk', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Bulk', 'Regular', 'Regular', 'Regular', 'Regular', 'Bulk', 'Regular', 'Regular', 'Bulk', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Bulk', 'Regular', 'Regular', 'Regular', 'Bulk', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Bulk', 'Regular', 'Regular', 'Regular', 'Bulk', 'Bulk', 'Bulk', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Bulk', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular', 'Regular'

In [30]:
# Create a 2D NumPy array
table = df[["Unit_Price", "Quantity", "Discount_Pct", "Revenue"]].to_numpy()

# Check the size of the array
print("Rows and columns:", table.shape)

# Show the first 3 rows
print("First 3 rows:")
print(table[:3])

# Show Revenue from the first 5 orders
print("Revenue of first 5 orders:")
print(table[:5, 3])

# Average of each column
print("Average of each column:")
print(table.mean(axis=0))

Rows and columns: (100, 4)
First 3 rows:
[[2.800e+02 1.000e+00 0.000e+00 2.800e+02]
 [3.000e+02 1.000e+00 1.000e-01 2.700e+02]
 [3.100e+02 1.000e+00 5.000e-02 2.945e+02]]
Revenue of first 5 orders:
[280.  270.  294.5 144.  480. ]
Average of each column:
[2.69800e+02 1.73000e+00 6.50000e-02 4.36525e+02]


In [33]:
# Missing values: Rating has 3 blank cells in Excel

rating = df["Rating"].to_numpy()

print("Average with missing values:", np.mean(rating))
print("Average ignoring missing:", round(np.nanmean(rating), 2))
print("Missing ratings:", np.isnan(rating).sum())

Average with missing values: nan
Average ignoring missing: 3.96
Missing ratings: 3


---
# Section 6: Pandas Series and DataFrame (8 min)

**Pandas** is the main tool for data analysis.

- **Series** = one column (values + labels called the *index*)
- **DataFrame** = a full table (many Series side by side), just like an Excel sheet

### 📖 Definitions
| Term | Definition | Starbucks example |
|---|---|---|
| **Pandas** | A library for data analysis. The name comes from *Panel Data*. | `import pandas as pd` |
| **Series** | A **one-dimensional** labelled array. It is one column of a table. | `df["Revenue"]` |
| **DataFrame** | A **two-dimensional** table with labelled rows and columns. Each column is a Series. | The 100 × 15 sales table |
| **Index** | The row labels of a Series or DataFrame. | 0, 1, 2... or Order IDs |
| **Column / Variable** | One field of the data. | `Price`, `Quantity` |
| **Row / Record** | One observation in the data. | One customer order |
| **`.head()` / `.tail()`** | Shows the first / last few rows. | Quick preview |
| **`.info()`** | Shows column names, data types and missing values. | Data quality check |
| **`.describe()`** | Shows summary statistics (mean, min, max...). | Average quantity |
| **`.value_counts()`** | Counts how many times each value appears. | Orders per city |
| **Missing value** | An empty cell, shown as `NaN`. Detect with `.isnull()`. | Blank Rating |

In [34]:
# Create a Series of product prices

price_series = pd.Series(
    [250, 270, 300, 350],
    index=["Cappuccino", "Latte", "Cold Brew", "Frappuccino"]
)

print(price_series)

# Increase all prices by 10%
print("After 10% price increase:")
print(price_series * 1.10)

Cappuccino     250
Latte          270
Cold Brew      300
Frappuccino    350
dtype: int64
After 10% price increase:
Cappuccino     275.0
Latte          297.0
Cold Brew      330.0
Frappuccino    385.0
dtype: float64


In [35]:
# A DataFrame column is a Series

rev = df["Revenue"]

print("Average revenue:", round(rev.mean(), 2))
print("Highest revenue:", rev.max())

# Count orders by city
print("Orders by city:")
print(df["City"].value_counts())

# Count orders by product
print("Orders by product:")
print(df["Product"].value_counts())

Average revenue: 436.52
Highest revenue: 2050.0
Orders by city:
City
Bengaluru    25
Mumbai       25
Hyderabad    20
Delhi        16
Pune         14
Name: count, dtype: int64
Orders by product:
Product
Latte               18
Cold Brew           15
Veg Sandwich        15
Cappuccino          13
Blueberry Muffin    13
Americano           10
Frappuccino         10
Masala Chai          6
Name: count, dtype: int64


In [36]:
# Create a small DataFrame
# Think of it like a small Excel table

mini = pd.DataFrame({

    # Product names
    "Product": ["Latte", "Cold Brew", "Masala Chai"],

    # Product categories
    "Category": ["Hot Coffee", "Cold Coffee", "Tea"],

    # Price of each product
    "Base_Price": [270, 300, 200]
})

# Display the DataFrame
print(mini)

       Product     Category  Base_Price
0        Latte   Hot Coffee         270
1    Cold Brew  Cold Coffee         300
2  Masala Chai          Tea         200


In [37]:
# DATAFRAME loaded from Excel: always begin with these 6 checks
print("1) Shape   :", df.shape)
print("2) Columns :", list(df.columns))
print("\n3) First 3 rows:")
print(df.head(3))
print("\n4) Last 3 rows:")
print(df.tail(3))
print("\n5) Info (data types and missing values):")
df.info()

1) Shape   : (100, 15)
2) Columns : ['OrderID', 'Date', 'City', 'StoreID', 'Product', 'Category', 'Size', 'Unit_Price', 'Quantity', 'Discount_Pct', 'Revenue', 'Payment_Mode', 'Customer_Type', 'Order_Channel', 'Rating']

3) First 3 rows:
   OrderID       Date       City StoreID     Product     Category    Size  \
0  SBX1001 2026-07-01  Hyderabad     S07  Cappuccino   Hot Coffee  Grande   
1  SBX1002 2026-07-01  Bengaluru     S05   Cold Brew  Cold Coffee    Tall   
2  SBX1003 2026-07-01     Mumbai     S03  Cappuccino   Hot Coffee   Venti   

   Unit_Price  Quantity  Discount_Pct  Revenue Payment_Mode Customer_Type  \
0         280         1          0.00    280.0          UPI           New   
1         300         1          0.10    270.0          UPI        Member   
2         310         1          0.05    294.5          UPI     Returning   

  Order_Channel  Rating  
0      In-store     4.0  
1      In-store     4.0  
2      In-store     2.0  

4) Last 3 rows:
    OrderID       Date  

In [38]:
# 6) Summary statistics of the numeric columns
print(df.describe().round(2))

print("\nMissing values per column:")
print(df.isnull().sum()[df.isnull().sum() > 0])      # only show columns with missing data
print("Unique cities:", df["City"].nunique(), "| Unique products:", df["Product"].nunique())

                      Date  Unit_Price  Quantity  Discount_Pct  Revenue  \
count                  100      100.00    100.00        100.00   100.00   
mean   2026-08-14 23:45:36      269.80      1.73          0.06   436.52   
min    2026-07-01 00:00:00      180.00      1.00          0.00   144.00   
25%    2026-07-19 18:00:00      240.00      1.00          0.00   242.25   
50%    2026-08-17 12:00:00      270.00      1.00          0.05   351.00   
75%    2026-09-07 00:00:00      300.00      2.00          0.10   567.25   
max    2026-09-30 00:00:00      410.00      5.00          0.20  2050.00   
std                    NaN       59.44      0.97          0.07   279.63   

       Rating  
count   97.00  
mean     3.96  
min      1.00  
25%      4.00  
50%      4.00  
75%      5.00  
max      5.00  
std      1.03  

Missing values per column:
Rating    3
dtype: int64
Unique cities: 5 | Unique products: 8


In [39]:
# Create a new column for discount amount
df["Discount_Amount"] = df["Unit_Price"] * df["Quantity"] * df["Discount_Pct"]

# Create a new column for month
df["Month"] = df["Date"].dt.month_name()

# Show the new columns
print(df[["OrderID", "Date", "Month", "Revenue", "Discount_Amount"]].head())

# Calculate total discount given
print("Total discount: Rs", round(df["Discount_Amount"].sum(), 2))

   OrderID       Date Month  Revenue  Discount_Amount
0  SBX1001 2026-07-01  July    280.0              0.0
1  SBX1002 2026-07-01  July    270.0             30.0
2  SBX1003 2026-07-01  July    294.5             15.5
3  SBX1004 2026-07-02  July    144.0             36.0
4  SBX1005 2026-07-04  July    480.0            120.0
Total discount: Rs 2447.5


---
# Section 7: Indexing and Selection (8 min)

How do we pick the rows and columns we need?

| Method | Meaning | Example |
|---|---|---|
| `df["col"]` | one column | `df["Revenue"]` |
| `df[["a","b"]]` | many columns | `df[["City","Revenue"]]` |
| `.iloc[ ]` | by **position** (numbers) | `df.iloc[0:3]` |
| `.loc[ ]` | by **label** (names) | `df.loc[0, "City"]` |
| Boolean filter | by **condition** | `df[df["Quantity"] > 2]` |

### 📖 Definitions
| Term | Definition | Starbucks example |
|---|---|---|
| **Indexing** | Picking a specific element, row or column using its label or position. | `df["Revenue"]` |
| **Selection** | Choosing a subset of data (rows and/or columns). | Only Pune rows |
| **`.loc[ ]`** | Selects by **label** (names). End of slice **is** included. | `df.loc[0:2, "City"]` |
| **`.iloc[ ]`** | Selects by **integer position**. End of slice is **not** included. | `df.iloc[0:3]` |
| **Boolean filtering** | Selecting rows where a condition is `True`. | `df[df["Revenue"] > 800]` |
| **`&` , `\|`** | AND / OR to combine conditions. Each condition must be inside brackets. | Mumbai **and** Cold Coffee |
| **`.isin()`** | Checks if values belong to a given list. | UPI or Wallet |
| **`.sort_values()` / `.nlargest()`** | Sort rows / get the top N rows. | Top 5 orders |
| **`.query()`** | Filter using a text condition. | `df.query("City == 'Pune'")` |
| **`set_index()`** | Makes one or more columns the row labels. | OrderID as index |
| **MultiIndex** | An index with more than one level. | (City, Product) |

In [40]:
# Select columns
print(df["Product"].head(3))
print()
print(df[["OrderID", "City", "Product", "Revenue"]].head(3))

0    Cappuccino
1     Cold Brew
2    Cappuccino
Name: Product, dtype: object

   OrderID       City     Product  Revenue
0  SBX1001  Hyderabad  Cappuccino    280.0
1  SBX1002  Bengaluru   Cold Brew    270.0
2  SBX1003     Mumbai  Cappuccino    294.5


In [43]:
# iloc = select by position

print("First 3 rows and 3 columns:")
print(df.iloc[0:3, 0:3])


# loc = select by label

print("City in row 4:")
print(df.loc[4, "City"])


print("OrderID, Product and Revenue for rows 0 to 2:")
print(df.loc[0:2, ["OrderID", "Product", "Revenue"]])

First 3 rows and 3 columns:
   OrderID       Date       City
0  SBX1001 2026-07-01  Hyderabad
1  SBX1002 2026-07-01  Bengaluru
2  SBX1003 2026-07-01     Mumbai
City in row 4:
Delhi
OrderID, Product and Revenue for rows 0 to 2:
   OrderID     Product  Revenue
0  SBX1001  Cappuccino    280.0
1  SBX1002   Cold Brew    270.0
2  SBX1003  Cappuccino    294.5


In [45]:
# Filtering data using conditions

# Show Pune orders
print("Pune orders:")
print(df[df["City"] == "Pune"])


# Show orders with revenue above Rs 800
print("Orders with revenue above Rs 800:")
print(df[df["Revenue"] > 800])

Pune orders:
    OrderID       Date  City StoreID           Product     Category     Size  \
9   SBX1010 2026-07-06  Pune     S09             Latte   Hot Coffee     Tall   
19  SBX1020 2026-07-13  Pune     S10             Latte   Hot Coffee     Tall   
32  SBX1033 2026-07-30  Pune     S10  Blueberry Muffin         Food  Regular   
44  SBX1045 2026-08-10  Pune     S09        Cappuccino   Hot Coffee     Tall   
53  SBX1054 2026-08-21  Pune     S09             Latte   Hot Coffee     Tall   
54  SBX1055 2026-08-21  Pune     S10         Cold Brew  Cold Coffee   Grande   
59  SBX1060 2026-08-24  Pune     S10       Frappuccino  Cold Coffee     Tall   
62  SBX1063 2026-08-26  Pune     S09         Cold Brew  Cold Coffee     Tall   
63  SBX1064 2026-08-26  Pune     S10      Veg Sandwich         Food  Regular   
75  SBX1076 2026-09-07  Pune     S10       Masala Chai          Tea     Tall   
88  SBX1089 2026-09-22  Pune     S10  Blueberry Muffin         Food  Regular   
91  SBX1092 2026-09-25  Pun

In [46]:
# Combining conditions

# 1. Mumbai AND Cold Coffee
print("Mumbai and Cold Coffee:")
print(df[(df["City"] == "Mumbai") &
         (df["Category"] == "Cold Coffee")])


# 2. UPI OR Wallet
print("Paid by UPI or Wallet:")
print(df[(df["Payment_Mode"] == "UPI") |
         (df["Payment_Mode"] == "Wallet")])


# 3. September AND Member AND 20% discount
print("September, Member, 20% discount:")
print(df[(df["Month"] == "September") &
         (df["Customer_Type"] == "Member") &
         (df["Discount_Pct"] == 0.20)])

Mumbai and Cold Coffee:
    OrderID       Date    City StoreID      Product     Category    Size  \
24  SBX1025 2026-07-19  Mumbai     S03    Cold Brew  Cold Coffee  Grande   
28  SBX1029 2026-07-23  Mumbai     S04  Frappuccino  Cold Coffee   Venti   
48  SBX1049 2026-08-17  Mumbai     S04    Cold Brew  Cold Coffee    Tall   
50  SBX1051 2026-08-18  Mumbai     S03    Cold Brew  Cold Coffee    Tall   
51  SBX1052 2026-08-18  Mumbai     S04  Frappuccino  Cold Coffee   Venti   
56  SBX1057 2026-08-22  Mumbai     S03  Frappuccino  Cold Coffee   Venti   
73  SBX1074 2026-09-07  Mumbai     S04    Cold Brew  Cold Coffee    Tall   
99  SBX1100 2026-09-30  Mumbai     S03  Frappuccino  Cold Coffee   Venti   

    Unit_Price  Quantity  Discount_Pct  Revenue Payment_Mode Customer_Type  \
24         330         1          0.15    280.5       Wallet           New   
28         410         1          0.05    389.5         Cash     Returning   
48         300         1          0.05    285.0         C

In [47]:
# Sorting data by Revenue

print("Top 5 orders by revenue:")

df_sorted = df.sort_values("Revenue", ascending=False)

print(df_sorted.head(5))

Top 5 orders by revenue:
    OrderID       Date       City StoreID           Product     Category  \
51  SBX1052 2026-08-18     Mumbai     S04       Frappuccino  Cold Coffee   
8   SBX1009 2026-07-05     Mumbai     S04             Latte   Hot Coffee   
98  SBX1099 2026-09-29  Bengaluru     S06             Latte   Hot Coffee   
92  SBX1093 2026-09-26  Bengaluru     S06        Cappuccino   Hot Coffee   
88  SBX1089 2026-09-22       Pune     S10  Blueberry Muffin         Food   

       Size  Unit_Price  Quantity  Discount_Pct  Revenue Payment_Mode  \
51    Venti         410         5          0.00   2050.0          UPI   
8      Tall         270         5          0.05   1282.5         Cash   
98    Venti         330         3          0.00    990.0          UPI   
92    Venti         310         3          0.00    930.0         Card   
88  Regular         180         5          0.00    900.0          UPI   

   Customer_Type Order_Channel  Rating  Discount_Amount      Month  
51     Ret

In [50]:
# Make OrderID the index
# Use City and Product as indexes

multi = df.set_index(["City", "Product"])

# Show Mumbai orders
print("Mumbai orders:")
print(multi.loc["Mumbai"])


# Show Mumbai + Latte orders
print("Mumbai Latte orders:")
print(multi.loc[("Mumbai", "Latte")])

Mumbai orders:
                  OrderID       Date StoreID     Category     Size  \
Product                                                              
Cappuccino        SBX1003 2026-07-01     S03   Hot Coffee    Venti   
Latte             SBX1009 2026-07-05     S04   Hot Coffee     Tall   
Americano         SBX1012 2026-07-07     S03   Hot Coffee   Grande   
Blueberry Muffin  SBX1013 2026-07-07     S03         Food  Regular   
Masala Chai       SBX1017 2026-07-11     S04          Tea     Tall   
Blueberry Muffin  SBX1018 2026-07-12     S03         Food  Regular   
Cold Brew         SBX1025 2026-07-19     S03  Cold Coffee   Grande   
Frappuccino       SBX1029 2026-07-23     S04  Cold Coffee    Venti   
Cappuccino        SBX1036 2026-08-01     S04   Hot Coffee    Venti   
Masala Chai       SBX1039 2026-08-03     S03          Tea   Grande   
Veg Sandwich      SBX1043 2026-08-10     S04         Food  Regular   
Cold Brew         SBX1049 2026-08-17     S04  Cold Coffee     Tall   
Cold 

/tmp/ipykernel_985/2222363836.py:13: PerformanceWarning: indexing past lexsort depth may impact performance.
  print(multi.loc[("Mumbai", "Latte")])


In [51]:
# MultiIndex: two index levels (City, Product)
multi = df.set_index(["City", "Product"]).sort_index()

print("All Mumbai rows (first 5):")
print(multi.loc["Mumbai"][["Quantity", "Revenue"]].head())

print("\nMumbai + Latte:")
print(multi.loc[("Mumbai", "Latte"), ["OrderID", "Quantity", "Revenue"]])

All Mumbai rows (first 5):
                  Quantity  Revenue
Product                            
Americano                1    225.0
Americano                2    440.0
Blueberry Muffin         1    180.0
Blueberry Muffin         1    144.0
Blueberry Muffin         3    540.0

Mumbai + Latte:
                OrderID  Quantity  Revenue
City   Product                            
Mumbai Latte    SBX1009         5   1282.5
       Latte    SBX1067         3    810.0
       Latte    SBX1072         1    300.0
       Latte    SBX1087         2    513.0
       Latte    SBX1088         1    216.0


---
# Section 8: Basic Data Structures for Organising Business Datasets (10 min)

Real companies keep data in **several tables**, not one big table. Our Excel file follows this design:

| Table (Excel sheet) | Type | Contains |
|---|---|---|
| `Products` | Master table | Product, Category, Base_Price, **Cost_Price** |
| `Stores` | Master table | StoreID, Store_Name, City, **Region**, **Store_Type** |
| `Sales_Data` | Transaction table | One row per order |

We join them with `merge` (like VLOOKUP) and summarise with `groupby` and `pivot_table`.

### 📖 Definitions
| Term | Definition | Starbucks example |
|---|---|---|
| **Dataset** | A collection of related data, usually in table form. | The Excel file |
| **Master table** | A table that stores stable reference information. | `Products`, `Stores` sheets |
| **Transaction table** | A table that records business events (sales, orders). | `Sales_Data` sheet |
| **Key / ID column** | A common column used to link two tables. | `StoreID`, `Product` |
| **Merge (Join)** | Combining two tables using a common key column. Similar to VLOOKUP in Excel. | Add Cost_Price and Region to sales |
| **GroupBy** | Splitting data into groups and applying a calculation to each group. | Revenue for each city |
| **Aggregation (`agg`)** | Summarising each group with sum, mean, count, etc. | Orders, revenue, avg rating per region |
| **Pivot table** | A summary table where one column becomes rows and another becomes columns. Same as Excel Pivot Table. | City × Category revenue |
| **Margins** | Row and column totals added to a pivot table. | TOTAL row |
| **Revenue** | Money earned from sales = Price × Quantity × (1 - Discount). | Rs 300 × 2 × 0.8 = Rs 480 |
| **Profit** | Revenue minus cost. | Revenue - (Cost_Price × Quantity) |
| **Profit margin** | Profit as a percentage of revenue. | Profit / Revenue × 100 |

In [ ]:
print("PRODUCTS (master):")
print(products_df)
print("\nSTORES (master):")
print(stores_df)

In [ ]:
# MERGE: bring Cost_Price (from Products) and Region/Store details (from Stores) into the sales table
full = (df.merge(products_df[["Product", "Cost_Price"]], on="Product", how="left")
          .merge(stores_df[["StoreID", "Store_Name", "Region", "Store_Type"]], on="StoreID", how="left"))

print("Rows before:", len(df), "| Rows after:", len(full), "(must stay equal)")
print("Columns before:", df.shape[1], "| Columns after:", full.shape[1])

# New business measures
full["Total_Cost"] = full["Cost_Price"] * full["Quantity"]
full["Profit"]     = full["Revenue"] - full["Total_Cost"]

print(full[["OrderID", "Store_Name", "Region", "Product", "Revenue", "Total_Cost", "Profit"]].head())

In [ ]:
# GROUPBY: split into groups, then calculate for each group
print("Revenue and profit by CITY:")
by_city = full.groupby("City")[["Revenue", "Profit"]].sum().round(2).sort_values("Revenue", ascending=False)
print(by_city)

print("\nRevenue by PRODUCT CATEGORY:")
by_category = full.groupby("Category")["Revenue"].sum().round(2).sort_values(ascending=False)
print(by_category)

In [ ]:
# Many measures at once with agg()
by_region = full.groupby("Region").agg(
    Orders        = ("OrderID", "count"),
    Total_Revenue = ("Revenue", "sum"),
    Avg_Order     = ("Revenue", "mean"),
    Total_Profit  = ("Profit", "sum"),
    Avg_Rating    = ("Rating", "mean"),
).round(2)
print(by_region)

print("\nTop 5 stores by revenue:")
print(full.groupby("Store_Name")["Revenue"].sum().round(2).sort_values(ascending=False).head(5))

In [ ]:
# Monthly trend (order the months correctly)
by_month = (full.groupby("Month")["Revenue"].sum().round(2)
                .reindex(["July", "August", "September"]))
print("Revenue by month:")
print(by_month)

print("\nPayment mode share (%):")
print((full["Payment_Mode"].value_counts(normalize=True) * 100).round(1))

print("\nAverage order value by customer type:")
print(full.groupby("Customer_Type")["Revenue"].mean().round(2))

In [ ]:
# PIVOT TABLE: City (rows) x Category (columns), like an Excel Pivot Table
pivot = full.pivot_table(values="Revenue", index="City", columns="Category",
                         aggfunc="sum", fill_value=0, margins=True, margins_name="TOTAL").round(0)
print(pivot)

print("\nOrders count: Order_Channel x Payment_Mode")
print(full.pivot_table(values="OrderID", index="Order_Channel", columns="Payment_Mode",
                       aggfunc="count", fill_value=0))

In [ ]:
# Business insights in one go
margin = (full.groupby("Category")["Profit"].sum() / full.groupby("Category")["Revenue"].sum() * 100).round(1)
print("Profit margin % by category:")
print(margin.sort_values(ascending=False))

best_store = full.groupby("Store_Name")["Profit"].sum().idxmax()
best_prod  = full.groupby("Product")["Profit"].sum().idxmax()
low_rating = full.groupby("Product")["Rating"].mean().idxmin()

print("\nMost profitable store   :", best_store)
print("Most profitable product :", best_prod)
print("Lowest-rated product    :", low_rating)

In [ ]:
# Save the summary tables into a NEW Excel file (3 sheets)
with pd.ExcelWriter("Starbucks_Summary.xlsx") as writer:
    by_city.to_excel(writer, sheet_name="By_City")
    by_region.to_excel(writer, sheet_name="By_Region")
    pivot.to_excel(writer, sheet_name="City_x_Category")

print("Saved: Starbucks_Summary.xlsx")

In [8]:
!find /content -name "Lectures1_3.ipynb"
!ls /content/your-repository

ls: cannot access '/content/your-repository': No such file or directory


---
# Lecture Summary

| Need | Use | Example from our data |
|---|---|---|
| Changeable sequence | `list` | Menu, list of revenues |
| Fixed record | `tuple` | Store details |
| Unique values, fast search | `set` | Unique cities, payment modes |
| Key → value lookup | `dict` | Product → price, OrderID → order |
| Fast numeric maths | NumPy `array` | Revenue = price × qty × (1 - discount) |
| One labelled column | Pandas `Series` | `df["Revenue"]` |
| Business table | Pandas `DataFrame` | The 100 × 15 sales table |
| Select data | `loc`, `iloc`, filters | Mumbai orders above Rs 800 |
| Combine tables | `merge` | Sales + Products + Stores |
| Summarise | `groupby`, `pivot_table` | Revenue by city and category |

# Practice Questions (for students)
1. Create a **set** of products sold in Delhi and a set sold in Pune. Which products are sold only in Delhi?
2. Using **NumPy**, find how many orders have a discount of more than 10%, and their total revenue.
3. Using **Pandas**, show all orders from Hyderabad paid by Cash or Card with Quantity 2 or more.
4. Using `.loc` and `set_index("OrderID")`, display the City and Revenue of orders SBX1020 to SBX1025.
5. Using `full`, find the average Rating for each Store_Type (Mall, High Street, Tech Park).
6. Using `pivot_table`, build Region × Customer_Type with the sum of Profit. Which cell is the highest?
7. **Challenge:** Which Month and Product combination earned the highest revenue?